### PR TITLE
✨  Introduce edge subnets to support AWS Local Zones

### PR DESCRIPTION
**Pull request by mtulio**
**What type of PR is this?**

/kind feature
/kind api-change

**What this PR does / why we need it**:

This PR implements support of managed subnets and carrier gateway for AWS Local Zones. Feature request #4874 .

There API is changed to introduce the following fields:
- `SubnetSpec.ZoneType`: representation of subnet's zone type
- `SubnetSpec.ParentZoneName`: representation of subnet's parent zone name (an availability zone in the Region which the edge zone is tied)

The subnets in AWS Local Zones are not eligible to create core components for the cluster, like NAT Gateway, Control Plane nodes, and Network Load Balancers, keeping compatibility with existing flow when edge subnets are added.

To create subnets in edge zones, the subnet must be added for each zone you want to create the subnet in `NetworkSpec.Subnets`. For example to create subnets in Local Zone `us-east-1-nyc-1a`, set:
```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AWSCluster
metadata:
  name: aws-cluster-localzone
spec:
  region: us-east-1
  networkSpec:
    vpc:
      cidrBlock: "10.0.0.0/20"
    subnets:
    # regular zones (availability zones. Eg us-east-1a, us-east-1b...)
    [...]
    # Subnets in Local Zones of New York location (public and private)
    - availabilityZone: us-east-1-nyc-1a
      cidrBlock: "10.0.128.0/25"
      id: "cluster-subnet-private-us-east-1-nyc-1a"
      isPublic: false
    - availabilityZone: us-east-1-nyc-1a
      cidrBlock: "10.0.128.128/25"
      id: "cluster-subnet-public-us-east-1-nyc-1a"
      isPublic: true
```

**Which issue(s) this PR fixes**: Relates to (not fixes completely) #4874

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Add support to provision subnets on AWS Local Zone infrastructure.
```

### CODE REVIEW

**Comment by mtulio** [comment]
/assign @vincepri @nrb 
cc @r4f4 @patrickdillon

**Comment by mtulio** [comment]
Hey @vincepri @nrb would you mind stamp `ok-to-test`? I am in the final phase of this PR to wrap up the Local Zones proposal. OpenShift CI is progressing/provisioning with this version (https://github.com/openshift/installer/pull/8173).

File: pkg/cloud/services/network/natgateways.go [hunk 1]
```diff
@@ -341,3 +341,39 @@ func (s *Service) getNatGatewayForSubnet(sn *infrav1.SubnetSpec) (string, error)
...
+	if sn.ParentZoneName != nil {
+		if gws, ok := azGateways[aws.StringValue(sn.ParentZoneName)]; ok && len(gws) > 0 {
+			return gws[0], nil
+		}
+	}
+
+	// Get the first public subnet's nat gateway available
+	for zone, gws := range azGateways {
+		if len(gws[0]) > 0 {
+			s.scope.Debug(fmt.Sprintf("Assigning route table ID %s to zone %s from zone %s\n", gws[0], sn.AvailabilityZone, zone))
```
Comment (r4f4) [inline pkg/cloud/services/network/natgateways.go:373]: Nit:
```suggestion
			s.scope.Debug("Assigning route table", "table ID", gws[0], "source zone", zone, "target zone", sn.AvailabilityZone)
```

**Review by r4f4 — COMMENTED** [review]
LGTM so far. Left a few nitpick comments, feel free to ignore.

**Comment by mtulio** [comment]
> LGTM so far. Left a few nitpick comments, feel free to ignore.

Thanks @r4f4 . Feedback addressed. 👍🏽 

I also removed some references of Wavelength Zones which will be addressed as "edge zone" in another PR.

**Comment by mtulio** [comment]
cc @r4f4 @patrickdillon @vincepri  this is ready for review, please take a look? I have some specific questions commented in the code, and some room to improve the e2e, but in general the Local Zones feature for managed and unmanaged network/subnets in Local Zones is implemented.

File: api/v1beta1/network_types.go [hunk 2]
```diff
@@ -276,6 +276,24 @@ type SubnetSpec struct {
...
+	//
+	// Zone types local-zone or wavelength-zone is not selected to automatically create
+	// resources like: Nat Gateway, Network Load Balancers, control plane or compute nodes.
+	//
+	// When local-zone, the public subnets will be associated to the public route table,
+	// the private subnets uses the zone in the region to create route tables re-using
+	// Nat Gateways (preferred from parent zone, the zone type availability-zone in the region,
+	// or first available).
+	// +optional
+	ZoneType *string `json:"zoneType,omitempty"`
```
Comment (vincepri) [inline api/v1beta1/network_types.go:292]: This should probably be an enum and validated accordingly using kubebuilder tags

**Comment by mtulio** [comment]
Converting to draft to address the PR review and unit tests failing due API conversions.

File: pkg/cloud/services/network/natgateways.go [hunk 3]
```diff
@@ -341,3 +341,39 @@ func (s *Service) getNatGatewayForSubnet(sn *infrav1.SubnetSpec) (string, error)
 
 	return "", errors.Errorf("no nat gateways available in %q for private subnet %q, current state: %+v", sn.AvailabilityZone, sn.GetResourceID(), azGateways)
 }
+
+func (s *Service) findNatGatewayForEdgeSubnet(sn *infrav1.SubnetSpec) (string, error) {
```
Comment (mtulio) [inline pkg/cloud/services/network/natgateways.go:345]: TODO(mtulio) introduce unit tests for this function.

File: api/v1beta2/network_types.go [hunk 4]
```diff
@@ -535,14 +581,40 @@ func (s Subnets) GetUniqueZones() []string {
...
+		if zone == nil {
+			return fmt.Errorf("unable to get zone information for zone %v", sub.AvailabilityZone)
+		}
+		if zone.ZoneType != nil {
+			sub.ZoneType = newZoneType(zone.ZoneType)
+		}
+		if zone.ParentZoneName != nil {
+			sub.ParentZoneName = zone.ParentZoneName
+		}
+	}
```
Comment (mtulio) [inline api/v1beta2/network_types.go:614]: TODO(mtulio): create `SetZoneInfo` in the SubnetSpec scope or remove from `Subnets` updating the reconcile to update zone info by zone, it would decrease the number of API calls to one by createSubnet, instead one by reconcliate loop.
Example: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4901/files#diff-1c519200705fb576b64e7fb271f7f683a2e6efadd54abd2a70b243e5069490e9R510-R532

**Comment by rvanderp3** [comment]
> /test pull-cluster-api-provider-aws-test

It appears the [test may have run out of quota for VPCs](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-test/1775323852826480640#1:build-log.txt%3A2291).  I'm re-running to see if this was a transient condition.

**Comment by nrb** [comment]
[Still failing on quota](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-test/1775522734072139776#1:build-log.txt%3A2288)

**Comment by rvanderp3** [comment]
> [Still failing on quota](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-test/1775522734072139776#1:build-log.txt%3A2288)

I actually think this is a red herring, we're actually failing on the conversion webhook.  I'm working up a fix now.

**Comment by rvanderp3** [comment]
@mtulio I think we need to update https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/api/v1beta1/awscluster_conversion.go#L105-L116 to 
```
	// Restore SubnetSpec.ResourceID, SubnetSpec.ParentZoneName, and SubnetSpec.ZoneType fields, if any.
	for _, subnet := range restored.Spec.NetworkSpec.Subnets {
		for i, dstSubnet := range dst.Spec.NetworkSpec.Subnets {
			if dstSubnet.ID == subnet.ID {
				if len(subnet.ResourceID) > 0 {
					dstSubnet.ResourceID = subnet.ResourceID
				}				
				if subnet.ParentZoneName != nil {
					dstSubnet.ParentZoneName = subnet.ParentZoneName
				}
				if subnet.ZoneType != nil {
					dstSubnet.ZoneType = (subnet.ZoneType)
				}
				dstSubnet.DeepCopyInto(&dst.Spec.NetworkSpec.Subnets[i])
			}
		}
	}
```

We aren't restoring the new fields.

**Comment by mtulio** [comment]
@damdo unit tests added, and everything is passing now.

File: pkg/cloud/services/network/natgateways.go [hunk 5]
```diff
@@ -341,3 +341,39 @@ func (s *Service) getNatGatewayForSubnet(sn *infrav1.SubnetSpec) (string, error)
...
+	if gws, ok := azGateways[sn.AvailabilityZone]; ok && len(gws) > 0 {
+		return gws[0], nil
+	}
+
+	// Check if the parent zone public subnet has nat gateway
+	if sn.ParentZoneName != nil {
+		if gws, ok := azGateways[aws.StringValue(sn.ParentZoneName)]; ok && len(gws) > 0 {
+			return gws[0], nil
+		}
+	}
```
Comment (rvanderp3) [inline pkg/cloud/services/network/natgateways.go:368]: @mtulio i have a couple of questions on this block:
- should we error out if we don't find a gateway associated with a ParentZoneName?
- I dont see where azGateways gets initialized with a key for the ParentZoneName.  Am I missing that?

Comment (mtulio) [reply pkg/cloud/services/network/natgateways.go:368]: Hey @rvanderp3 :
> should we error out if we don't find a gateway associated with a ParentZoneName?
This is part of the design/fallback strategy. If there is not GW for the ParentZone, it will fallback to the first gateway found (line 374)

>  I dont see where azGateways gets initialized with a key for the ParentZoneName. Am I missing that?
`azGAteways` are populated in the line [356][356] with NatGateway IDs for all public subnets.

[]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882/files#diff-23c065949155c8a08cd81009d31c08a2d86909abe59c908a374d11c9a641281eR356

File: pkg/cloud/services/network/subnets.go [hunk 6]
```diff
@@ -151,7 +151,10 @@ func (s *Service) reconcileSubnets() error {
 				// We may not have a permission to tag unmanaged subnets.
 				// When tagging unmanaged subnet fails, record an event and proceed.
 				record.Warnf(s.scope.InfraCluster(), "FailedTagSubnet", "Failed tagging unmanaged Subnet %q: %v", existingSubnet.GetResourceID(), err)
-				break
+				// Review Bug(mtulio): if proceed, out of the loop, here will create subnets without AvailabilityZone, after the future
```
Comment (mtulio) [inline pkg/cloud/services/network/subnets.go:154]: TODO(mtulio) review this comment before merge

File: pkg/cloud/services/network/natgateways_test.go [hunk 7]
```diff
@@ -40,6 +40,161 @@ const (
...
+				CidrBlock:      "10.0.10.0/24",
+				ParentZoneName: aws.String("us-east-1d"),
+			},
+			expect: *knownSubnets[3].NatGatewayID,
+		},
+		{
+			name: "return nat gateway ID for the input subnet with a parent zone name and availability zone",
+			input: infrav1.SubnetSpec{
+				CidrBlock:        "10.0.10.0/24",
+				AvailabilityZone: "us-east-1c",
```
Comment (mtulio) [inline pkg/cloud/services/network/natgateways_test.go:122]: I would change the zones with parent to Local Zone name. Example:
```suggestion
				AvailabilityZone: "us-east-1-bos-1a",
```

File: pkg/cloud/services/network/natgateways_test.go [hunk 8]
```diff
@@ -40,6 +40,161 @@ const (
...
+			IsPublic:         true,
+			NatGatewayID:     aws.String("nat-1234567890abcdef1"),
+		},
+		{
+			ID:               "subnet-4",
+			AvailabilityZone: "us-east-1d",
+			CidrBlock:        "10.0.40.0/24",
+			IsPublic:         true,
+			NatGatewayID:     aws.String("nat-1234567890abcdef2"),
+			ParentZoneName:   aws.String("zone-subnet-4"),
```
Comment (mtulio) [inline pkg/cloud/services/network/natgateways_test.go:71]: As we are using valid zone names, let's use it to the ParentZoneName too
```suggestion
			ParentZoneName:   aws.String("us-east-1"),
```

File: pkg/cloud/services/network/natgateways_test.go [hunk 9]
```diff
@@ -40,6 +40,157 @@ const (
...
+			input: infrav1.SubnetSpec{
+				ID:               "subnet-7",
+				CidrBlock:        "10.0.10.0/24",
+				AvailabilityZone: "us-east-1-nyc-1a",
+				ParentZoneName:   aws.String("us-east-1d"),
+			},
+			expect: *knownSubnets[3].NatGatewayID,
+		},
+		{
+			name: "return nat gateway ID for the input subnet with an invalid parent zone name and availability zone",
```
Comment (mtulio) [inline pkg/cloud/services/network/natgateways_test.go:125]: the case is ok, what about to improve the name for something like this:
> "return nat gateway ID for the first valid subnet when an invalid parent zone name and availability zone"

**Comment by mtulio** [comment]
Converting to draft to cherry pick changes for edge zones from Wavelength PR.

**Comment by mtulio** [comment]
Can't find the to

> @mtulio: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-cluster-api-provider-aws-e2e	[b5e14d5](https://github.com/kubernetes-sigs/cluster-api-provider-aws/commit/b5e14d5b6c596739d49a4f527d9b19ff76cb4df7)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-e2e/1778881489476259840)	false	`/test pull-cluster-api-provider-aws-e2e`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-aws&pr=4882). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Amtulio). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues?q=is:issue+is:open) when you hit one in your PR.

Can't determine the root cause. Re-testing

~~~
  INFO: Error getting pod capi-kubeadm-bootstrap-system/capi-kubeadm-bootstrap-controller-manager-5fff6c8b98-rcl85, container manager: Get "https://self-hosted-zjqo26-apiserver-1855160729.us-west-2.elb.amazonaws.com:6443/api/v1/namespaces/capi-kubeadm-bootstrap-system/pods/capi-kubeadm-bootstrap-controller-manager-5fff6c8b98-rcl85": stream error: stream ID 2339; INTERNAL_ERROR; received from peer

~~~

/test pull-cluster-api-provider-aws-e2e

**Comment by mtulio** [comment]
For reviewers: this PR is generally ready for review.

I am still investigating if the timeouts in the e2e is related with this PR, i will update when I have some information (non-blocking for review).

File: api/v1beta2/network_types.go [hunk 10]
```diff
@@ -37,6 +41,11 @@ const (
 	DefaultAPIServerHealthThresholdCount = 5
 	// DefaultAPIServerUnhealthThresholdCount the API server unhealthy check threshold count.
 	DefaultAPIServerUnhealthThresholdCount = 3
+
+	// ZoneTypeAvailabilityZone defines the regular AWS zones in the Region.
+	ZoneTypeAvailabilityZone = "availability-zone"
+	// ZoneTypeLocalZone defines the AWS zone type in Local Zone infrastructure.
+	ZoneTypeLocalZone = "local-zone"
```
Comment (r4f4) [inline api/v1beta2/network_types.go:48]:
```suggestion
	// ZoneTypeAvailabilityZone defines the regular AWS zones in the Region.
	ZoneTypeAvailabilityZone ZoneType = "availability-zone"
	// ZoneTypeLocalZone defines the AWS zone type in Local Zone infrastructure.
	ZoneTypeLocalZone ZoneType = "local-zone"
```

File: api/v1beta2/network_types.go [hunk 11]
```diff
@@ -446,6 +475,45 @@ func (s *SubnetSpec) String() string {
...
+// IsEdge returns the true when the subnet is created in the edge zone,
+// Local Zones.
+func (s *SubnetSpec) IsEdge() bool {
+	if s.ZoneType == nil {
+		return false
+	}
+	if s.ZoneType.Equal(ZoneType(ZoneTypeLocalZone)) {
+		return true
+	}
+	return false
```
Comment (r4f4) [inline api/v1beta2/network_types.go:487]:
```suggestion
    return s.ZoneType != nil && *s.ZoneType == ZoneTypeLocalZone
```

File: api/v1beta2/network_types.go [hunk 12]
```diff
@@ -446,6 +475,45 @@ func (s *SubnetSpec) String() string {
...
+
+	zone := zoneInfo(s.AvailabilityZone)
+	if zone == nil {
+		if len(s.AvailabilityZone) > 0 {
+			return fmt.Errorf("unable to update zone information for subnet '%v' and zone '%v'", s.ID, s.AvailabilityZone)
+		}
+		return fmt.Errorf("unable to update zone information for subnet '%v'", s.ID)
+	}
+	if zone.ZoneType != nil {
+		s.ZoneType = ptr.To(ZoneType(*zone.ZoneType))
```
Comment (r4f4) [inline api/v1beta2/network_types.go:509]: Maybe I'm just being paranoid, but I feel like it's safer to save the value to a local var before taking its address:
```suggestion
        var zt := ZoneType(*zone.ZoneType)
		s.ZoneType = ptr.To(zt)
```

File: api/v1beta2/network_types.go [hunk 13]
```diff
@@ -501,48 +585,70 @@ func (s Subnets) FindEqual(spec *SubnetSpec) *SubnetSpec {
 }
 
 // FilterPrivate returns a slice containing all subnets marked as private.
-func (s Subnets) FilterPrivate() (res Subnets) {
+func (s Subnets) FilterPrivate() Subnets {
```
Comment (r4f4) [inline api/v1beta2/network_types.go:588]: Any special reason to remove the named return?

File: api/v1beta2/network_types.go [hunk 14]
```diff
@@ -501,48 +585,70 @@ func (s Subnets) FindEqual(spec *SubnetSpec) *SubnetSpec {
...
 			res = append(res, x)
 		}
 	}
-	return
+	return res
 }
 
 // FilterPublic returns a slice containing all subnets marked as public.
-func (s Subnets) FilterPublic() (res Subnets) {
+func (s Subnets) FilterPublic() Subnets {
```
Comment (r4f4) [inline api/v1beta2/network_types.go:603]: ditto

File: api/v1beta2/network_types.go [hunk 15]
```diff
@@ -501,48 +585,70 @@ func (s Subnets) FindEqual(spec *SubnetSpec) *SubnetSpec {
...
 			res = append(res, x)
 		}
 	}
-	return
+	return res
 }
 
 // FilterByZone returns a slice containing all subnets that live in the availability zone specified.
-func (s Subnets) FilterByZone(zone string) (res Subnets) {
+func (s Subnets) FilterByZone(zone string) Subnets {
```
Comment (r4f4) [inline api/v1beta2/network_types.go:618]: ditto

File: api/v1beta2/network_types.go [hunk 16]
```diff
@@ -501,48 +585,70 @@ func (s Subnets) FindEqual(spec *SubnetSpec) *SubnetSpec {
...
 		}
 	}
 	return zones
 }
 
+// SetZoneInfo updates the subnets with zone information.
+func (s Subnets) SetZoneInfo(zones []*ec2.AvailabilityZone) error {
+	for i := range s {
+		err := s[i].SetZoneInfo(zones)
+		if err != nil {
```
Comment (r4f4) [inline api/v1beta2/network_types.go:645]:
```suggestion
		if err := s[i].SetZoneInfo(zones); err != nil {
```

File: api/v1beta2/network_types.go [hunk 17]
```diff
@@ -757,3 +863,16 @@ func (i *IngressRule) Equals(o *IngressRule) bool {
...
+
+// String returns the string representation for the zone type.
+func (z ZoneType) String() string {
+	return string(z)
+}
+
+// Equal compares two zone types.
+func (z ZoneType) Equal(other ZoneType) bool {
+	return z == other
+}
```
Comment (r4f4) [inline api/v1beta2/network_types.go:878]: Is this needed?

File: api/v1beta2/network_types.go (hunk 11 above)
Comment (r4f4) [reply api/v1beta2/network_types.go:487]: Then I wouldn't mind leaving it as it is. I forgot this was broken out from the future changes.

File: api/v1beta2/network_types.go (hunk 13 above)
Comment (mtulio) [reply api/v1beta2/network_types.go:588]: that's makes sense. I was writing some test cases to get empty scenarios, but just realized it wont more required. I am reverting this change adding more unit cases to test empty values.

File: api/v1beta2/network_types.go (hunk 17 above)
Comment (mtulio) [reply api/v1beta2/network_types.go]: Not for now. FUP https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882#discussion_r1566355658 : 
I am removing from this PR, it will introduce in WL when we'll need more complex check w/ more edge types. Thanks

File: api/v1beta2/network_types.go (hunk 12 above)
Comment (r4f4) [reply api/v1beta2/network_types.go:503]: Oh, I didn't mean safer in the sense of the value being `nil` but more in the sense of saving the addr of a temporary variable. I guess that doesn't make sense since it's basically doing `&(*zonetype)` which is `zonetype`.

**Comment by mtulio** [comment]
> @mtulio: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command

> pull-cluster-api-provider-aws-test	[848a1f9](https://github.com/kubernetes-sigs/cluster-api-provider-aws/commit/848a1f95af1f364e2501682811241b883b924a8c)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-test/1779995138043416576)	true	`/test pull-cluster-api-provider-aws-test`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-aws&pr=4882). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Amtulio). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues?q=is:issue+is:open) when you hit one in your PR.

Unit test[1] from package [2] was failing missing the new API calls to describe zones in the mock. Unit test fixed.

[1] `TestAWSManagedControlPlaneReconcilerIntegrationTests`
[2] `sigs.k8s.io/cluster-api-provider-aws/v2/controlplane/eks/controllers`

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Comment by mtulio** [comment]
> @mtulio: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name | Commit | Details | Required | Rerun command
> --- | --- | --- | --- | ---
> pull-cluster-api-provider-aws-e2e | c3d8c20f7283f698c59c4e22d08f9562ab996565 | [link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-e2e/1780049347019280384) | false | `/test pull-cluster-api-provider-aws-e2e`
> 
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-aws&pr=4882). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Amtulio). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues?q=is:issue+is:open) when you hit one in your PR.
> 
> <details>
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
> </details>
> 

It looks like there are some instability in the most recent versions to get passing in e2e. I am tracking three PRs[1][2][3], and looks like 1 and 2 are unhealthy with early timeouts, or related to upgrades.

The 3 is going well, which is two commits behind current the master version, where 1 and 2 are on top off. The 3 is also a superset of 2. Thoughts?

1) https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4849
2) https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882
3)
https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4901

**Comment by nrb** [comment]
> It seems there are some instability in the most recent versions to get passing in e2e. ... Thoughts?

I'm not sure what's happening here. Maybe we're running close to capacity? Maybe resources are deleting faster than we expect? Opened https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4937 to track as we do keep seeing timeouts with the v1beta1 to v1beta2 test on other PRs. If you see it on other PRs, tag it so we can gather more data.

It looks like [1] is passing CI now, which is another thing that makes me think this is a flake.

File: api/v1beta2/network_types.go [hunk 18]
```diff
@@ -430,6 +439,26 @@ type SubnetSpec struct {
...
 	Tags Tags `json:"tags,omitempty"`
+
+	// ZoneType defines a zone type for this subnet.
+	//
+	// The valid values are availability-zone, and local-zone.
+	//
+	// Zone types local-zone is not selected to automatically create
+	// resources like Nat Gateway, Network Load Balancers, control plane or compute nodes.
+	//
+	// When local-zone, the public subnets will be associated with the public route table,
```
Comment (nrb) [inline api/v1beta2/network_types.go:450]: When the `ZoneType` is `availability-zone`, what happens with the route table and subnets? They're just treated like normal ones in an AZ? Would be nice to document that explicitly.

File: api/v1beta2/network_types.go [hunk 19]
```diff
@@ -430,6 +439,26 @@ type SubnetSpec struct {
...
+	// the private subnets use the zone in the region to create route tables re-using
+	// NAT Gateways (preferred from the parent zone, the zone type availability-zone in the region,
+	// or first table available).
+	//
+	// +kubebuilder:validation:Enum=availability-zone;local-zone
+	// +optional
+	ZoneType *ZoneType `json:"zoneType,omitempty"`
+
+	// ParentZoneName defines a parent zone name of the zone that the subnet is created.
+	// +optional
```
Comment (nrb) [inline api/v1beta2/network_types.go:460]: Is this field only valid if `ZoneType == ZoneTypeLocalZone`?

Comment (mtulio) [reply api/v1beta2/network_types.go:460]: Yes (and Wavelenght in the follow up PRs), this field will automatically filled by reconcile from zone describe API. Do you think we need to create some CEL here or just improve the documentation?

File: api/v1beta2/network_types.go (hunk 18 above)
Comment (mtulio) [reply api/v1beta2/network_types.go:450]: Yes, tables and subnets for regular zones (`availability-zone`) is not affected, keeping the regular flow. Thanks for the feedback, I will update the field documentation.

**Comment by mtulio** [comment]
> I have a couple questions that could be answered as comments in the code, but I don't think I have any objections.

Hey @nrb , feedback addressed. Let me know what do you think. Thanks!

**Comment by nrb** [comment]
@mtulio I think it's good from side!

/lgtm

**Comment by nrb** [comment]
Just needs updated files.

`generated files are out of date, run make generate`

**Comment by mtulio** [comment]
> Just needs updated files.
> 
> `generated files are out of date, run make generate`

interesting, I ran it before [latest change](pull-cluster-api-provider-aws-verify). Let me check if it was something cached.

/test pull-cluster-api-provider-aws-verify

**Comment by r4f4** [comment]
> > Just needs updated files.
> > `generated files are out of date, run make generate`
> 
> interesting, I ran it before [latest change](pull-cluster-api-provider-aws-verify), I ran now resulting in no changes. Let me check if it was something cached.

@mtulio you need to rebase on top of current master. https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4897 bumped the version of `controller-gen` which is changing the results.

**Comment by mtulio** [comment]
> > > Just needs updated files.
> > > `generated files are out of date, run make generate`
> > 
> > 
> > interesting, I ran it before [latest change](pull-cluster-api-provider-aws-verify), I ran now resulting in no changes. Let me check if it was something cached.
> 
> @mtulio you need to rebase on top of current master. #4897 bumped the version of `controller-gen` which is changing the results.

Thanks @r4f4 . Rebased from current master, `verify` is now passing. Triggering e2e's with rebased version:

/test pull-cluster-api-provider-aws-e2e
/test pull-cluster-api-provider-aws-e2e-eks

**Comment by mtulio** [comment]
Timeout.
/test pull-cluster-api-provider-aws-e2e

**Comment by mtulio** [comment]
> > > > Just needs updated files.
> > > > `generated files are out of date, run make generate`
> > > 
> > > 
> > > interesting, I ran it before [latest change](pull-cluster-api-provider-aws-verify), I ran now resulting in no changes. Let me check if it was something cached.
> > 
> > 
> > @mtulio you need to rebase on top of current master. #4897 bumped the version of `controller-gen` which is changing the results.
> 
> Thanks @r4f4 . Rebased from current master, `verify` is now passing. Triggering e2e's with rebased version:
> 
> /test pull-cluster-api-provider-aws-e2e /test pull-cluster-api-provider-aws-e2e-eks

@nrb @richardcase  okay, e2e* are passing after rebase.

**Comment by richardcase** [comment]
As mentioned by @mtulio the e2e passed. For posterity:
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-e2e/1781467444657786880
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4882/pull-cluster-api-provider-aws-e2e-eks/1781401927121113088

**Comment by richardcase** [comment]
@mtulio - would you be able to squash your commits? It doesn't have to be squashed into 1 if you want to tell a story by the separate commits.

**Comment by mtulio** [comment]
> @mtulio - would you be able to squash your commits? It doesn't have to be squashed into 1 if you want to tell a story by the separate commits.

Hey @richardcase , sure! Thanks for the feedback. I just squashed into 3 commits. Let me know if it looks better. Thanks!

**Review by richardcase — COMMENTED** [review]
Looks good. Just the one comment around adding some CEL validation (or even a validation to the webhook).

File: api/v1beta2/network_types.go [hunk 20]
```diff
@@ -430,6 +439,39 @@ type SubnetSpec struct {
...
+	ZoneType *ZoneType `json:"zoneType,omitempty"`
+
+	// ParentZoneName is the zone name where the current subnet's zone is tied when
+	// the zone is a Local Zone.
+	//
+	// The subnets in Local Zone locations consume the ParentZoneName to determine the correct
+	// private route table to egress traffic to the internet.
+	//
+	// +optional
+	ParentZoneName *string `json:"parentZoneName,omitempty"`
```
[the remark this answers is not in the document]
Comment (mtulio) [reply api/v1beta2/network_types.go:552]: > Or use validation in the webhook.
yeah, when thinking in the follow up PRs (Wavelength), it seems to be better to create a validation in the webhook. Let me do it instead of CEL.

**Comment by mtulio** [comment]
Hey @nrb @richardcase [validation added][val] in the subnet's webhook to prevent setting empty `ParentZoneName` when `ZoneType` is `local-zone`. Let me know what do you think. Thanks

[val]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882/files#diff-ee5083c9ded05729c760b5c8d9a4ac50a896c995c497d629ac2da13cc2ef2df4R251-R255

File: api/v1beta2/network_types.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (mtulio) [reply api/v1beta2/network_types.go:552]: Webhook added

File: api/v1beta2/awscluster_webhook.go [hunk 21]
```diff
@@ -248,6 +248,11 @@ func (r *AWSCluster) validateNetwork() field.ErrorList {
 		if subnet.IsIPv6 || subnet.IPv6CidrBlock != "" {
 			allErrs = append(allErrs, field.Invalid(field.NewPath("subnets"), r.Spec.NetworkSpec.Subnets, "IPv6 cannot be used with unmanaged clusters at this time."))
 		}
+		if subnet.ZoneType != nil && subnet.IsEdge() {
+			if subnet.ParentZoneName == nil {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("subnets"), r.Spec.NetworkSpec.Subnets, "ParentZoneName must be set when ZoneType is 'local-zone'."))
+			}
+		}
```
Comment (damdo) [inline api/v1beta2/awscluster_webhook.go:255]: Note for the reviewer:
```
// IsEdge returns the true when the subnet is created in the edge zone,
// Local Zones.
```

**Review by damdo — COMMENTED** [review]
/lgtm

**Comment by richardcase** [comment]
We should follow-up on a CAPA approach to CEL vs validation webhooks. But lets not let that block this.

So from my side:

/approve
